New CAIP profile: interoperable address binary specification (#350)
ChainAgnostic/CAIPs: CAIPs/caip-350.md

---
caip: 350
title: Binary serialization of blockchain IDs and addresses
author: Defi Wonderland (@defi-wonderland), Teddy (@0xteddybear), Joxes (@Joxess), Racu (@0xRacoon), Skeletor Spaceman (@0xskeletor-spaceman), TiTi (@0xtiti), Gori (@0xGorilla), Ardy (@0xArdy), Onizuka (@onizuka-wl)
discussions-to: https://ethereum-magicians.org/t/erc-7930-interoperable-addresses/23365
status: Draft
type: Standard
created: 2025-04-18
requires: 2
---

## Simple Summary
A CAIP profile for every chain namespace to describe how to serialize their addresses and chain references into the Interoperable Address format described in [ERC-7930].

## Abstract
Every chain defines their own address types & also a way to identify networks within that chain's ecosystem.
[ERC-7930] is an address format for (address, chain) pairs supporting any chain, and also including metadata on the chain & addresses' type so no information needs to be communicated out of band.

> This CAIP namespace aims to be a living & mutable repository of serialization methods for all chains to uniformly conform to Interoperable Addresses.

Reviewer: In CAIP context, the word "namespace" has different meaning. It means namespaces such as EIP-155, BIP-122, etc. See https://namespaces.chainagnostic.org/

Author: you're right, I meant 'profile'. Will edit this, thanks!

Author: addressed!

## Motivation
Standards like CAIP-2 and CAIP-10 are simple text representations of addresses & chain namespaces/references, but they do not address:
- Binary representation of addresses and/or chain namespace+reference: Relevant & desired for on-chain usage.

> - Canonicity: CAIP-10, being a generic text format, leaves the canonization and serialization of ASCII address to on-chain addresses to each namespace's profile; unless a given namespace's profile specifies such logic, the standard itself makes no universal guarantees on a blockchain account (_target address_ in ERC-7930 parlance) having only one CAIP-10 representation, which makes them less useful as dictionary keys.

Reviewer 2: ```suggestion
- Canonicity: CAIP-10, being a generic text format, leaves the canonization and serialization of ASCII address to on-chain addresses to each namespace's profile; unless a given namespace's profile specifies such logic, the standard itself makes no universal guarantees on a blockchain account (_target address_ in ERC-7930 parlance) having only one CAIP-10 representation; for namespaces where such canonicity is neither inherent or specified by the CASA profile, duplicate entries can be created (e.g. when used as dictionary keys).
```

> - Canonicity: CAIP-10, being a generic text format, leaves the canonization and serialization of ASCII address to on-chain addresses to each namespace's profile; unless a given namespace's profile specifies such logic, the standard itself makes no universal guarantees on a blockchain account (_target address_ in ERC-7930 parlance) having only one CAIP-10 representation, which makes them less useful as dictionary keys.

Reviewer: This means we should fix CAIP-10 instead of introducing new standard

Reviewer 2: i'm not sure it's fixable in general, given how divergent addressing and account models are across namespaces; each namespace needs to be profiled down and specify its own canonicity behaviors, in my opinion.

Author: update: this standard already defines the 'Extra Considerations' sections to deal with this, but added the possibility of address/chain collisions as a 'Security Consideration' of ERC-7930

- Completeness: both formats have limits on the length of their identifiers, which are reasonable for human-readability but often identifiers have to drop meaningful information to conform to those length requirements. While that information might be easy to look up in the context of wallet software, doing so within a smart contract would not be possible.
- Succinctness: Text formats necessarily have to incur encoding overhead compared to binary ones, causing relative informational inefficiency.

## Specification

> The purpose of every supporting profile is be to be able to deterministically and unambiguously convert between:

Reviewer 2: ```suggestion
The purpose of each namespace's profile is to specify deterministic and unambiguous conversions between the following format-pairs:
```

> - Customary address text formats of the ecosystem (which may also be described in its [CAIP-10] profile) and the Interoperable Address text representation defined in [ERC-7930]

Reviewer 2: I think it would be more precise to refer to all these as "ASCII" formats than "text", since AFAIK all of them are the latter and the former could be confused with, e.g. "seed phrases" or more human-readable/speakable formats.

Author: I'm not sure the definition of text formats (such as CAIP-10, and EIP-55 to a lesser extent since it mentions ascii/utf8 but I'm not sure how normative it intends to be about it) specify the text encoding to be used, only that the text is meant to contain such and such latin/punctuation _characters_). Since they are meant to be displayed to humans, the underlying encoding could be considered an implementation detail, and describing them as ASCII could potentially be incorrect.

I do think the document could benefit from clearly differentiating between resolved (potentially ERC-7828 Interoperable Names) and un-resolved (ERC-7930 Interoperable Names) names. Do you think there are better words for those concepts though? I can't seem to be able to come up with something both clear and concise

- Customary chain reference text formats of the ecosystem (which may also be described in [CAIP-2]) and the text representation defined in [ERC-7930]

> - [ERC-7930]'s Interoperable Address -> [ERC-7930]'s Interoperable Name and back of the ecosystem's chain references

Reviewer 2: ```suggestion
- From [ERC-7930]'s Interoperable Address -> [ERC-7930]'s Interoperable Name to the ecosystem's chain references and back
```

Reviewer 2: the others are just pairs, but this specifies "to and back" -- is there a different in the expectation of losslessness?

Author: no, I wrote it like that because ...I thought it was cleaner natural language? but losslessness expectations are the same. Rewrote to match the items above

> - [ERC-7930]'s Interoperable Address -> [ERC-7930]'s Interoperable Name and back of the ecosystem's addresses *for all address formats of said ecosystem*

Reviewer 2: ```suggestion
- [ERC-7930]'s Interoperable Address -> [ERC-7930]'s Interoperable Name to *all address formats of said ecosystem* and back
```

And must also define the binary id of the namespace itself, akin to [ERC-7930]'s `ChainType`

> For this, every namespace MUST define the following sections in their definition of their CAIP-350 profile:

Reviewer 2: ```suggestion
For this, every namespace MUST define specify all of the above in a CAIP-350 profile to maximize interoperability and review. A [template for these profiles](https://github.com/ChainAgnostic/namespaces/blob/main/_template/caip350.md) is defined in the Namespaces registry.
```

And move the template to the URL above by opening a separate PR [over in namespaces](https://github.com/ChainAgnostic/namespaces/pulls)

<!-- instructions for converting from the text representation to the binary one -->

#### Examples

### Extra considerations
<!-- Anything that is particular to this namespace and of interest to users, such as not being able to satisfy canonicity requirements -->
```

## Rationale
The main alternative to this standard would've been to define all formats & conversions in [ERC-7930], turning it into a Living ERC, which was not ideal since:
- Living standards are not a usual thing in the Ethereum ecosystem: the only other Living ERC is [EIP-1]
- It would be desireable to finalize the definition of the format itself while allowing the specification of every chain's serialization to be defined as needed.
- It would have placed all the editorial (and presumably most of the specification) work on the [ERC-7930] authors, while the CASA paradigm allows better distribution of work by leaving the definition of the profile for every chain namespace to the parties of its ecosystem interested in using Interoperable Addresses.

## Test Cases
Not applicable since this does not define a normative interface, instead being a meta-specification of normative interfaces.

## Security Considerations

> It is possible that a CAIP-000 profile for a given chain namespace is not able to guarantee requirements of canonicity which are a desired property of Interoperable Addresses due to characteristics of the chain namespace, and systems relying on that canonicity may exhibit unexpected behavior as a result. When this is the case, it should be noted in the `Extra Considerations` section.

Reviewer 2: ```suggestion
It is possible that a CAIP-350 profile for a given chain namespace is not able to guarantee requirements of canonicity which are a desired property of Interoperable Addresses due to characteristics of the chain namespace, and systems relying on that canonicity may exhibit unexpected behavior as a result. When this is the case, it should be noted in the `Extra Considerations` section.
```

i wish we had written a section like this for CAIP-10 before it went to final 🙈

---
caip: 350
title: Binary serialization of blockchain IDs and addresses
author: Defi Wonderland (@defi-wonderland), Teddy (@0xteddybear), Joxes (@Joxess), Racu (@0xRacoon), Skeletor Spaceman (@0xskeletor-spaceman), TiTi (@0xtiti), Gori (@0xGorilla), Ardy (@0xArdy), Onizuka (@onizuka-wl)
discussions-to: https://ethereum-magicians.org/t/erc-7930-interoperable-addresses/23365
status: Draft
type: Standard
created: 2025-04-18
requires: 2
---

## Simple Summary

> A CAIP profile for every chain namespace to describe how to serialize their addresses and chain references into the Interoperable Address format described in [ERC-7930].

Reviewer 2: ```suggestion
This standard, when profiled down for each namespace, describes a unified and general purpose method for serializing chain-specified addresses into the Interoperable Address format specified canonically in [ERC-7930].
```

## Abstract

> Every chain defines their own address types & also a way to identify networks within that chain's ecosystem.

Reviewer 2: ```suggestion
Every namespaces defines their own address types & also a way to identify networks within that namespace's ecosystem.
```

> [ERC-7930] is an address format for (address, chain) pairs supporting any chain, and also including metadata on the chain & addresses' type so no information needs to be communicated out of band.

Reviewer 2: ```suggestion
[ERC-7930] is an address format for (address, chain) pairs supporting any chain, and also including metadata on the chain & addresses' type in a self-describing way, reducing the need to pass out-of-band metadata or context.
```

This CAIP aims to append a living & mutable repository of chain-specific address serialization methods to the CASA system in the form of its namespace-specific profiles, enabling all current and future chains to uniformly conform to Interoperable Addresses.

## Motivation
Standards like CAIP-2 and CAIP-10 are simple text representations of addresses & chain namespaces/references, but they do not address:
- Binary representation of addresses and/or chain namespace+reference: Relevant & desired for on-chain usage.

> - Canonicity: CAIP-10, being a generic text format, leaves the canonization and serialization of text address to on-chain addresses to each namespace's profile; unless a given namespace's profile specifies such logic, the standard itself makes no universal guarantees on a blockchain account (_target address_ in ERC-7930 parlance) having only one CAIP-10 representation; for namespaces where such canonicity is neither inherent or specified by the CASA profile, duplicate entries can be created (e.g. when used as dictionary keys).

Reviewer 2: ```suggestion
- Canonicity: CAIP-10, being a generic text format, leaves the canonization and serialization of text address to on-chain addresses to each namespace's profile. Unless a given namespace's profile specifies such logic, the standard itself makes no universal guarantees on a blockchain account (_target address_ in ERC-7930 parlance) having only one CAIP-10 representation. For namespaces where such canonicity is neither inherent nor specified by the CASA profile, duplicate entries can be created (e.g. when used as dictionary keys).
```

And must also define the binary id of the namespace itself, akin to [ERC-7930]'s `ChainType`

For this, every namespace MUST define specify all of the above in a CAIP-350 profile to maximize interoperability and review. A [template for these profiles](https://github.com/ChainAgnostic/namespaces/blob/main/_template/caip350.md) is defined in the Namespaces registry.

## Rationale
The main alternative to this standard would've been to define all formats & conversions in [ERC-7930], turning it into a Living ERC, which was not ideal since:
- Living standards are not a usual thing in the Ethereum ecosystem: the only other Living ERC is EIP-1
- It would be desireable to finalize the definition of the format itself while allowing the specification of every chain's serialization to be defined as needed.
- It would have placed all the editorial (and presumably most of the specification) work on the [ERC-7930] authors, while the CASA paradigm allows better distribution of work by leaving the definition of the profile for every chain namespace to the parties of its ecosystem interested in using Interoperable Addresses.

## Test Cases

> Not applicable since this does not define a normative interface, instead being a meta-specification of normative interfaces.

Reviewer 2: ```suggestion
Not applicable since this does not define a normative interface, instead being a meta-specification of normative interfaces.
Test cases SHOULD be included in each namespace profile.
```

## Security Considerations
It is possible that a CAIP-350 profile for a given chain namespace is not able to guarantee requirements of canonicity which are a desired property of Interoperable Addresses due to characteristics of the chain namespace, and systems relying on that canonicity may exhibit unexpected behavior as a result. When this is the case, it should be noted in the `Extra Considerations` section.

## Backwards Compatibility

> This standard actively seeks to be exhaustive in its backwards compatibility with CAIP-2 and CAIP-10.

Reviewer 2: I'm still not clear on how this works with Completeness, and whether a CAIP-350 implementation is allowed to produce or required to accept a CAIP-2 or CAIP-10 in cases where these truncate the information preserved in binary or in the novel format!

---
caip: 350

> title: Binary serialization of blockchain IDs and addresses

Reviewer 3: ```suggestion
title: Binary Serialization of Blockchain IDs and Addresses
```

author: Defi Wonderland (@defi-wonderland), Teddy (@0xteddybear), Joxes (@Joxess), Racu (@0xRacoon), Skeletor Spaceman (@0xskeletor-spaceman), TiTi (@0xtiti), Gori (@0xGorilla), Ardy (@0xArdy), Onizuka (@onizuka-wl)
discussions-to: https://ethereum-magicians.org/t/erc-7930-interoperable-addresses/23365
status: Draft
type: Standard
created: 2025-04-18
requires: 2
---

## Simple Summary

This standard, when profiled down for each namespace, describes a unified and general purpose method for serializing chain-specified addresses into the Interoperable Address format specified in [ERC-7930].

## Abstract

> Every namespace defines their own address types & also a way to identify networks within that namespace's ecosystem.

Reviewer 3: ```suggestion
Every namespace defines their own address types and also a way to identify networks within that namespace's ecosystem.
```

[ERC-7930] is an address format for (address, chain) pairs supporting any chain, and also including metadata on the chain & addresses' type in a self-describing way, reducing the need to communicate metadata or context out-of-band.
This CAIP aims to append a living & mutable repository of chain-specific address serialization methods to the CASA system in the form of its namespace-specific profiles, enabling all current and future chains to uniformly conform to Interoperable Addresses.

## Motivation

> Standards like CAIP-2 and CAIP-10 are simple text representations of addresses & chain namespaces/references, but they do not address:

Reviewer 3: ```suggestion
Standards like CAIP-2 and CAIP-10 are simple text representations of addresses & chain namespaces and references, but they do not address:
```

- Binary representation of addresses and/or chain namespace+reference: Relevant & desired for on-chain usage.
- Canonicity: CAIP-10, being a generic text format, leaves the canonization and serialization of text address to on-chain addresses to each namespace's profile. Unless a given namespace's profile specifies such logic, the standard itself makes no universal guarantees on a blockchain account (_target address_ in ERC-7930 parlance) having only one CAIP-10 representation. For namespaces where such canonicity is neither inherent nor specified by the CASA profile, duplicate entries can be created (e.g. when used as dictionary keys).

> - Completeness: both formats have limits on the length of their identifiers, which are reasonable for human-readability but often identifiers have to drop meaningful information to conform to those length requirements. While that information might be easy to look up in the context of wallet software, doing so within a smart contract would not be possible.

Reviewer 3: ```suggestion
- Completeness: Both formats have limits on the length of their identifiers, which are reasonable for human-readability, but often identifiers have to drop meaningful information to conform to those length requirements. While that information might be easy to look up in the context of wallet software, doing so within a smart contract would not be possible.
```

- It would have placed all the editorial (and presumably most of the specification) work on the [ERC-7930] authors, while the CASA paradigm allows better distribution of work by leaving the definition of the profile for every chain namespace to the parties of its ecosystem interested in using Interoperable Addresses.

## Test Cases

Not applicable since this does not define a normative interface, instead being a meta-specification of normative interfaces.

Test cases SHOULD be included in each namespace profile.

## Security Considerations

It is possible that a CAIP-350 profile for a given chain namespace is not able to guarantee requirements of canonicity which are a desired property of Interoperable Addresses due to characteristics of the chain namespace, and systems relying on that canonicity may exhibit unexpected behavior as a result. When this is the case, it should be noted in the `Extra Considerations` section.

## Backwards Compatibility

This standard actively seeks to be exhaustive in its backwards compatibility with CAIP-2 and CAIP-10.

> Namespace profiles SHOULD clarify which conversions to and from the aforementioned standards are:

Reviewer 2: ```suggestion
Namespace profiles SHOULD clarify which conversions to and from the aforementioned standards are:

```

> Namespaces are responsible over clarifying which conversions to and from the aforementioned standards are:

Reviewer 2: ```suggestion
Namespace profiles SHOULD clarify which conversions to and from the aforementioned standards are:
```
